spec/type_references_and_ownership.md

```diff
@@ -0,0 +1,331 @@
+    // If you're referencing a subnet: Microsoft.Network/virtualNetworks/{network}/subnets/{subnet}
+    // TODO: There are a ton of options for how we do this... some of the requiring more manual work than others... we should discuss this
+    Name string
+}
+```
+
+This approach seems to have the most flexibility and power. It allows easy reference to static Azure resources
+not managed by Kubernetes but also allows users to stay entirely within the Kubernetes ecosystem and refer to everything by
+its Kubernetes name.
+
+The primary downside of this approach is complexity - references aren't just a single string they're a complex
+object and the customer will need to choose which flavor they want each time they use a reference.
+
+#### Implementation
+There are two key implementation details for this approach:
+1. We must identify all resource references in the JSON schema and transform their type to `AugmentedKnownTypeReference`.
+   There is no specific marker which means: "This field is a reference" - most are called `id` but that's not a guarantee.
+   For example on the [VirtualMachineScaleSetIPConfigurationProperties](https://schema.management.azure.com/schemas/2019-07-01/Microsoft.Compute.json) 
+   the `subnet` field is of custom type `ApiEntityReference`, which has an `id` field where you put the ARM ID for a subnet.
+   This may require some manual work. One thing we can investigate doing long term
+   is see if there's a way to get teams to annotate "links" in their Swagger somehow...
+2. We have to be able to build the fully qualified ARM ID for any referenced resource.
+
+### How to represent ownership and dependent resources
+Many resources have owners but do not have an `id` field referring to their owner. Similarly, many resources with dependent resources may not
+have a field which refers to them.
+**We need to code generate a field which allows users of the CRDs to express ownership**.
+As discussed in [what ARM does](#What-ARM-does), this is the `resources` property in the ARM deployment templates.
+
+When we determine that a resource is a dependent resource of another resource kind, we will code-generate
+an `OwnerRef` (actual name TBD) property in the dependent resource `Spec`. This will also include an annotation 
+about the expected type of the resource.
+
+```go
+type SubnetSpec struct {
+    OwnerRef AugmentedKnownTypeReference `json:"ownerRef" group:"microsoft.network.infra.azure.com" kind:"VirtualNetwork"`
+    ...
+}
+```
+
+When users submit a dependent object we should validate that the provided owner reference is present.
+This can be accomplished by making the property required in the CRD.
+One major advantage of this approach is that the customer cannot really get the link type wrong, because we've autogenerated the expected
+group/kind information all names they supply must point to the right kind of resource. If they are using the `AzureReference` portion
```

Comment:
Why not just create an "unmanaged" resource for the Azure referenced resource. For example, the user could create a resource by specifying the only the ID for a `VirtualNetwork`.

```yaml
apiVersion: microsoft.network.infra.azure.com/v20191101
kind: VirtualNetwork
metadata:
  name: sample-virtualnetwork
spec:
  id: /subscriptions/1234/resourcegroups/foo/microsoft.network/virtualnetworks/bazz
```
Perhaps, the controller could fetch the data from the ID and populate the spec from that data. The K8s resource could be marked with an annotation which would keep the Azure resource from being deleted when the K8s resource is deleted.

The above solution would allow for importing Azure resources, simplifies the resource reference in K8s by not needing to create a `AzureReference` b/c everything will be a `KnowTypeReference`, and it provides a simple way to know what K8s need to clean up and what is an "unmanaged" resource.

Reply:
As discussed I'm sorta 50-50 on this. I think this approach has some advantages and some disadvantages over what I originally proposed. I've tried to cover those in the "Open Design Question - shape of Azure References" section of the document and have asked some other folks to weigh in to see what makes the most sense.